### PR TITLE
Tweak `parse_dot_suffix_expr`

### PR DESCRIPTION
**Pull request by nnethercote**
I find this function hard to understand, so I rewrote it.

r? @est31

### CODE REVIEW

**Comment by nnethercote** [comment]
Best reviewed one commit at a time.

File: compiler/rustc_parse/src/parser/expr.rs [hunk 1]
```diff
@@ -1008,41 +1009,40 @@ impl<'a> Parser<'a> {
             }
             token::Literal(token::Lit { kind: token::Float, symbol, suffix }) => {
                 Ok(match self.break_up_float(symbol, self.token.span) {
-                    // 1e2
```
Comment (est31) [inline compiler/rustc_parse/src/parser/expr.rs]: FYI I copied these comments from another location in this file. If you grep for "1.2 | 1.2e3" for example, you'll find a bunch of locations in this file, not all of them involving `DestructuredFloat`.

**Review by est31 — COMMENTED** [review]
r+ me except for that comment point. maybe the original form can be preserved for grepability somehow, and the additional info can be included behind it?

**Comment by nnethercote** [comment]
I reinstated the original comments as requested.

@bors r=est31

**Pull request merged**